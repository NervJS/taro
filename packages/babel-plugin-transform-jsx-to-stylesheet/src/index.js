const path = require('path')

const STYLE_SHEET_NAME = '_styleSheet'
const GET_STYLE_FUNC_NAME = '_getStyle'
const MERGE_STYLES_FUNC_NAME = '_mergeStyles'
const GET_CLS_NAME_FUNC_NAME = '_getClassName'
const NAME_SUFFIX = 'StyleSheet'
const cssSuffixs = ['.css', '.scss', '.sass', '.less', '.styl']
let styleNames = [] // css module 写法中 import 的 Identifier

module.exports = function ({types: t, template}) {
  const mergeStylesFunctionTemplate = template(`
function ${MERGE_STYLES_FUNC_NAME}() {
  var newTarget = {};

  for (var index = 0; index < arguments.length; index++) {
    var target = arguments[index];

    for (var key in target) {
      newTarget[key] = Object.assign(newTarget[key] || {}, target[key]);
    }
  }

  return newTarget;
}
  `)
  const getClassNameFunctionTemplate = template(`
function ${GET_CLS_NAME_FUNC_NAME}() {
  var className = [];
  var args = arguments[0];
  var type = Object.prototype.toString.call(args).slice(8, -1).toLowerCase();

  if (type === 'string') {
    args = args.trim();
    args && className.push(args);
  } else if (type === 'array') {
    args.forEach(function (cls) {
      cls = ${GET_CLS_NAME_FUNC_NAME}(cls).trim();
      cls && className.push(cls);
    });
  } else if (type === 'object') {
    for (var k in args) {
      k = k.trim();
      if (k && args.hasOwnProperty(k) && args[k]) {
        className.push(k);
      }
    }
  }

  return className.join(' ').trim();
}
  `)
  const getStyleFunctionTemplete = template(`
function ${GET_STYLE_FUNC_NAME}(classNameExpression) { 
  var className = ${GET_CLS_NAME_FUNC_NAME}(classNameExpression);
  var classNameArr = className.split(/\\s+/);

  var style = [];
  if (classNameArr.length === 1) {
    style.push(${STYLE_SHEET_NAME}[classNameArr[0].trim()]);
  } else {
      classNameArr.forEach(function(cls) {
      style.push(${STYLE_SHEET_NAME}[cls.trim()]);
    });
  }

  return style;
}
  `)

  const getClassNameFunctionAst = getClassNameFunctionTemplate()
  const mergeStylesFunctionAst = mergeStylesFunctionTemplate()
  const getStyleFunctionAst = getStyleFunctionTemplete()

  function getArrayExpression (value) {
    let expression
    let str

    if (!value || value.value === '') {
      // className
      // className=""
      return []
    } else if (value.type === 'JSXExpressionContainer' && value.expression && typeof value.expression.value !== 'string') {
      // className={{ container: true }}
      // className={['container wrapper', { scroll: false }]}
      return [t.callExpression(t.identifier(GET_STYLE_FUNC_NAME), [value.expression])]
    } else {
      // className="container"
      // className={'container'}
      str = (value.expression ? value.expression.value : value.value).trim()
    }

    return str === '' ? [] : str.split(/\s+/).map((className) => {
      return template(`${STYLE_SHEET_NAME}["${className}"]`)().expression
    })
  }

  function findLastImportIndex (body) {
    const bodyReverse = body.slice(0).reverse()
    let _index = 0

    bodyReverse.some((node, index) => {
      if (node.type === 'ImportDeclaration') {
        _index = body.length - index - 1
        return true
      }
      return false
    })

    return _index
  }

  return {
    visitor: {
      Program: {
        exit (astPath, {file}) {
          const node = astPath.node
          // const cssFileCount = file.get('cssFileCount')
          const injectGetStyle = file.get('injectGetStyle')
          const lastImportIndex = findLastImportIndex(node.body)
          let cssParamIdentifiers = file.get('cssParamIdentifiers')
          let callExpression

          if (cssParamIdentifiers) {
            // only one css file，由于样式文件合并，永远只有一个
            if (cssParamIdentifiers.length === 1) {
              callExpression = t.variableDeclaration('var', [t.variableDeclarator(t.identifier(STYLE_SHEET_NAME), cssParamIdentifiers[0])])
            } else if (cssParamIdentifiers.length > 1) {
              const objectAssignExpression = t.callExpression(t.identifier(MERGE_STYLES_FUNC_NAME), cssParamIdentifiers)
              callExpression = t.variableDeclaration('var', [t.variableDeclarator(t.identifier(STYLE_SHEET_NAME), objectAssignExpression)])
            }

            node.body.splice(lastImportIndex + 1, 0, callExpression)

            if (injectGetStyle) {
              node.body.splice(lastImportIndex + 2, 0, getClassNameFunctionAst)
              node.body.splice(lastImportIndex + 3, 0, getStyleFunctionAst)
            }
          }

          // 全部合并成一个，也就不需要 mergeStylesFunctionAst 了
          // if (cssFileCount > 1) {
          //   node.body.unshift(mergeStylesFunctionAst)
          // }

          // 将 css modules 的 styles 替换为 _styleSheet
          astPath.traverse({
            JSXOpeningElement (astPath) {
              astPath.traverse({
                JSXAttribute ({node}) {
                  if (node.name && node.name.name === 'style') {
                    astPath.traverse({
                      MemberExpression ({node}) {
                        if (node.object.type === 'Identifier' && styleNames.indexOf(node.object.name) > -1) {
                          node.object.name = STYLE_SHEET_NAME
                        }
                      }
                    })
                  }
                }
              })
            }
          })
          // 清除该 js 文件的 styleNames
          styleNames = []
        }
      },
      JSXOpeningElement ({container}, {file}) {
        const cssFileCount = file.get('cssFileCount') || 0
        if (cssFileCount < 1) {
          return
        }

        // Check if has "style"
        let hasStyleAttribute = false
        let styleAttribute
        let hasClassName = false
        let classNameAttribute

        const attributes = container.openingElement.attributes
        for (let i = 0; i < attributes.length; i++) {
          const name = attributes[i].name
          if (name) {
            if (!hasStyleAttribute) {
              hasStyleAttribute = name.name === 'style'
              styleAttribute = hasStyleAttribute && attributes[i]
            }

            if (!hasClassName) {
              hasClassName = name.name === 'className'
              classNameAttribute = hasClassName && attributes[i]
            }
          }
        }

        if (hasClassName) {
          // Remove origin className
          attributes.splice(attributes.indexOf(classNameAttribute), 1)

          if (
            classNameAttribute.value &&
            classNameAttribute.value.type === 'JSXExpressionContainer' &&
            typeof classNameAttribute.value.expression.value !== 'string' // not like className={'container'}
          ) {
            file.set('injectGetStyle', true)
          }

          const arrayExpression = getArrayExpression(classNameAttribute.value)

          if (arrayExpression.length === 0) {
            return
          }

          if (hasStyleAttribute && styleAttribute.value) {
            let expression = styleAttribute.value.expression
            let expressionType = expression.type

            // style={[styles.a, styles.b]} ArrayExpression
            if (expressionType === 'ArrayExpression') {
              expression.elements = arrayExpression.concat(expression.elements)
              // style={styles.a} MemberExpression
              // style={{ height: 100 }} ObjectExpression
              // style={{ ...custom }} ObjectExpression
              // style={custom} Identifier
              // style={getStyle()} CallExpression
              // style={this.props.useCustom ? custom : null} ConditionalExpression
              // style={custom || other} LogicalExpression
            } else {
              styleAttribute.value.expression = t.arrayExpression(arrayExpression.concat(expression))
            }
          } else {
            let expression = arrayExpression.length === 1 ? arrayExpression[0] : t.arrayExpression(arrayExpression)
            attributes.push(t.jSXAttribute(t.jSXIdentifier('style'), t.jSXExpressionContainer(expression)))
          }
        }
      },
      // 由于目前 js 引入的文件样式默认会全部合并，故进插入一个就好，其余的全部 remove
      ImportDeclaration (astPath, state) {
        const {file} = state
        const node = astPath.node
        const sourceValue = node.source.value
        const specifiers = node.specifiers
        const jsFilePath = state.opts.filePath // 传进来的文件的 filePath ,Babel6 file.opts.filaname 为unknown
        const extname = path.extname(sourceValue)
        const cssIndex = cssSuffixs.indexOf(extname)
        let cssFileCount = file.get('cssFileCount') || 0
        let cssParamIdentifiers = file.get('cssParamIdentifiers') || []

        if (cssIndex > -1) {
          // `import styles from './foo.css'` kind
          if (node.importKind === 'value' && specifiers.length > 0) {
            styleNames.push(specifiers[0].local.name)
          }
          // 第一个引入的样式文件
          if (cssFileCount === 0) {
            const cssFileBaseName = path.basename(jsFilePath, path.extname(jsFilePath))
            // 引入样式对应的变量名
            const styleSheetIdentifierValue = `${cssFileBaseName.replace(/-/g, '_') + NAME_SUFFIX}`
            const styleSheetIdentifierPath = `./${cssFileBaseName}_styles`
            const styleSheetIdentifier = t.identifier(styleSheetIdentifierValue)

            node.specifiers = [t.importDefaultSpecifier(styleSheetIdentifier)]
            node.source = t.stringLiteral(styleSheetIdentifierPath)
            cssParamIdentifiers.push(styleSheetIdentifier)
          } else {
            astPath.remove()
          }
          cssFileCount++

          file.set('cssParamIdentifiers', cssParamIdentifiers)
          file.set('cssFileCount', cssFileCount)
        }
      }
    }
  }
}
