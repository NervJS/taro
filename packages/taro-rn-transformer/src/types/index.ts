export const globalAny: any = global
export interface TransformType {
  src: string,
  filename: string,
  options: {
    nextTransformer: any,
    entry?: string,
    sourceRoot?: string,
    projectRoot: string,
    appName?: string,
    isEntryFile: (filename: string) => boolean,
    designWidth?: number,
    deviceRatio?: Record<number, number>
  }
}

export interface TransformPage {
  projectRoot: string,
  filename: string,
  sourceCode: string,
  sourceDir: string,
}

export interface TransformEntry {
  sourceDir: string,
  appName: string,
  projectRoot: string,
  filename: string,
  designWidth: number,
  deviceRatio: Record<number, number>,
  entryName: string
}

export interface AppConfig {
  pages: string[]
  subPackages?: SubPackage[]
  subpackages?: SubPackage[],
  designWidth: number,
  deviceRatio?: Record<number, unknown>,
  tabBar:Record<string, any>
}

interface SubPackage {
  /**
   * 分包根路径
   * - 注意：不能放在主包pages目录下
   */
  root: string
  /**
   * 分包路径下的所有页面配置
   */
  pages: string[]
}
