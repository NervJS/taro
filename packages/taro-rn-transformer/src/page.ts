import * as path from 'path'
import { TransformPage, globalAny } from './types/index'

export default function generatePage ({ sourceCode, filename, projectRoot, sourceDir }: TransformPage) {
  // 文件
  const extName = path.basename(filename).split('.')[0]
  const fileDir = path.dirname(filename)
  let result = sourceCode
  if (!(fileDir === sourceDir && extName === 'app')) { // 非入口文件
    const commonStyle = globalAny?.__taroCommonStyle || []
    if (commonStyle && commonStyle.length > 0) {
      const code: string[] = []
      const filePath = path.join(projectRoot, filename)
      commonStyle.forEach((item) => {
        let importStr = ''
        const relativePath = path.relative(path.dirname(filePath), item.path).replace(/\\/g, '/')
        const realPath = path.dirname(filePath) === path.dirname(item.path) ? `./${item.fileName}` : `${relativePath}`
        if (item.name) {
          importStr = `import ${item.name} from '${realPath}'`
        } else {
          importStr = `import '${realPath}'`
        }
        code.push(importStr)
      })
      result = code.join(';\n') + ';' + sourceCode
    }
  }
  return result
}
