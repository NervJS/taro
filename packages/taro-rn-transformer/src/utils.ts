import * as nodePath from 'path'
import * as fs from 'fs'
import * as parser from '@babel/parser'
import traverse from '@babel/traverse'
import { readConfig, resolveMainFilePath } from '@tarojs/helper'
import { globalAny } from './types/index'

const RN_CSS_EXT = ['.css', '.scss', '.sass', '.less', '.styl', '.stylus']

export function getConfigFilePath (filePath: string) {
  return resolveMainFilePath(`${filePath.replace(nodePath.extname(filePath), '')}.config`)
}

export function getConfigContent (path: string) {
  if (!path) return {}
  const fileConfigPath = getConfigFilePath(path)
  const content = readConfig(fileConfigPath)
  return content
}

export function getStyleCode (code: string, basePath: string) {
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: [
      'jsx',
      'typescript',
      'classProperties',
      'decorators-legacy'
    ]
  })

  const styleTypes = RN_CSS_EXT
  const styleSource: Record<string, string>[] = []
  traverse(ast, {
    ImportDeclaration (path) {
      const node = path.node
      const resource: string = node.source.value
      const ext = nodePath.extname(resource)
      // 是否是样式文件
      if (!styleTypes.includes(ext)) {
        return
      }
      let sourceName = ''
      if (node.specifiers.length) {
        sourceName = node.specifiers[0].local.name
      }
      const realPath = nodePath.resolve(basePath, resource)
      const fileName = nodePath.basename(realPath)
      styleSource.push({
        name: sourceName,
        path: realPath,
        fileName: fileName
      })
    }
  })
  return styleSource
}

export function isPageFile (file: string, sourceDir: string) {
  if ((/node_modules/.test(file)) || file.indexOf(sourceDir) === -1) return false
  const pagesList = globalAny.__taroAppPages || []
  const dirname = nodePath.dirname(file).replace(/\\/g, '/')
  const fileObj = nodePath.parse(file)
  const name = fileObj.name.split('.')[0]
  const filePath = `${dirname}/${name}`
  const filename = nodePath.basename(file).replace(nodePath.extname(file), '')
  return pagesList.includes(filePath) && !(filename.endsWith('.config'))
}

export function getFileContent (fileName: string) {
  let code = ''
  if (!fs.existsSync(fileName)) return code
  try {
    code = fs.readFileSync(fileName, 'utf-8').toString()
  } catch (error) {
    code = ''
  }
  return code
}

export function getCommonStyle (appPath: string, basePath: string) {
  let styles: Record<string, string>[] = []
  // 读取入口文件的内容
  const jsExt: string[] = ['tsx', 'ts', 'jsx', 'js']
  let codeStr = ''
  // 先读带rn后缀的
  for (let i = 0; i < jsExt.length; i++) {
    const rnfilePath = `${appPath}.rn.${jsExt[i]}`
    const rnFileContent: string = getFileContent(rnfilePath)
    if (!rnFileContent) {
      codeStr = rnFileContent
      break
    }
  }
  // 不带rn后缀的
  if (!codeStr) {
    for (let i = 0; i < jsExt.length; i++) {
      const filePath = `${appPath}.${jsExt[i]}`
      const fileContent: string = getFileContent(filePath)
      if (fileContent) {
        codeStr = fileContent
        break
      }
    }
  }
  if (!codeStr) return styles
  styles = getStyleCode(codeStr, basePath)
  return styles
}
