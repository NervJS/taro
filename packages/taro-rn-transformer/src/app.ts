import * as path from 'path'
import * as fs from 'fs'
import { camelCase } from 'lodash'
import { isEmptyObject } from '@tarojs/helper'
import { getConfigContent, getConfigFilePath } from './utils'
import { TransformEntry, AppConfig } from './types/index'

function getPagesResource (appPath: string, basePath: string, pathPrefix: string) {
  const importPages: string[] = []
  const screenPages: string[] = []
  const importConfigs: string[] = []
  const pages = getAppPages(appPath)

  pages.forEach(item => {
    const pagePath = item.startsWith('/') ? item : `/${item}`
    const screenName = camelCase(pagePath)
    const importScreen = `import ${screenName} from '.${pathPrefix}${pagePath}'`
    importPages.push(importScreen)
    screenPages.push(pagePath)
    const configFile = getConfigFilePath(path.join(basePath, pagePath))
    const screenConfigName = `${screenName}Config`
    if (fs.existsSync(configFile)) {
      importConfigs.push(`import ${screenConfigName} from '.${pathPrefix}${pagePath}.config'`)
    } else {
      importConfigs.push(`const ${screenConfigName} = {}`)
    }
  })
  return {
    screenPages,
    importPages,
    importConfigs
  }
}

function getPageScreen (pagePath: string) {
  const screen = camelCase(pagePath)
  const screenConfigName = `${screen}Config`
  return `{name:'${screen}',pagePath:'${pagePath}',component:createPageConfig(${screen},{...${screenConfigName},pagePath:'${pagePath}'})}`
}

function getAppConfig (appPath: string) {
  // 读取配置文件内容
  if (!appPath) {
    throw new Error('缺少 app 全局配置文件，请检查！')
  }
  const appConfig: AppConfig = getConfigContent(appPath)
  if (isEmptyObject(appConfig)) {
    throw new Error('缺少 app 全局配置，请检查！')
  }
  if (appConfig && (!appConfig.pages || !appConfig.pages.length)) {
    throw new Error('全局配置缺少 pages 字段，请检查！')
  }
  return appConfig
}

export function getAppPages (appPath: string) {
  const config = getAppConfig(appPath)
  const pages = config?.pages || []
  // 分包路由，也需要处理
  const subPackages = config.subPackages || config.subpackages || []
  subPackages.forEach(item => {
    const subRoot = item.root.endsWith('/') ? item.root : `${item.root}/`
    const subPages = item.pages
    subPages.forEach(itm => {
      pages.push(subRoot + itm)
    })
  })
  const res = pages.map(item => { return item.startsWith('/') ? item : `/${item}` })
  return res
}

export default function generateEntry ({
  filename,
  projectRoot,
  sourceDir,
  appName,
  entryName,
  designWidth,
  deviceRatio
}: TransformEntry) {
  // 文件后缀
  const filePath = path.basename(filename).replace(path.extname(filename), '')
  const basePath = path.join(projectRoot, sourceDir)
  const appPath = path.join(projectRoot, sourceDir, entryName)

  const appConfig = {
    designWidth,
    deviceRatio
  }
  const pathPrefix = filePath.indexOf(sourceDir) > -1 ? '' : `/${sourceDir}`
  const pages = getPagesResource(appPath, basePath, pathPrefix)
  const importPageList = pages.importPages.join(';')
  const importPageConfig = pages.importConfigs.join(';')
  const routeList = pages.screenPages
  const appComponentPath = `./${sourceDir}/${entryName}`

  const code = `import { AppRegistry } from 'react-native'
  import { createReactNativeApp, createPageConfig } from '@tarojs/runtime-rn'
  import Component from '${appComponentPath}'
  ${importPageList}
  ${`import AppComponentConfig from '${appComponentPath}.config';`}
  ${importPageConfig}

  const buildConfig = ${JSON.stringify(appConfig)}
  const config = { appConfig: { ...buildConfig, ...AppComponentConfig } }
  global.__taroAppConfig = config
  config['pageList'] = [${routeList.map(pageItem => getPageScreen(pageItem))}]
  AppRegistry.registerComponent('${appName}',() => createReactNativeApp(Component,config))
  `
  return code
}
