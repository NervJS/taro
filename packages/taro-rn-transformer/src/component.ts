import * as path from 'path'

import { globalAny, TransformPage } from './types/index'
import { transformLinaria } from './utils'

export default function componentLoader ({ sourceCode, filename, projectRoot, sourceDir }: TransformPage) {
  const filePath = path.join(projectRoot, filename)

  // 文件
  const extName = path.basename(filename).split('.')[0]
  const fileDir = path.dirname(filename)
  let result = sourceCode
  // 入口文件不加入全局样式 commonStyle
  if (!(fileDir === sourceDir && extName === 'app')) {
    const commonStyle = globalAny?.__taroCommonStyle || []
    if (commonStyle && commonStyle.length > 0) {
      const code: string[] = []
      commonStyle.forEach((item) => {
        let importStr = ''
        const relativePath = path.relative(path.dirname(filePath), item.path).replace(/\\/g, '/')
        const realPath = path.dirname(filePath) === path.dirname(item.path) ? `./${item.fileName}` : `${relativePath}`
        if (item.name) {
          importStr = `import ${item.name} from '${realPath}'`
        } else {
          importStr = `import '${realPath}'`
        }
        code.push(importStr)
      })
      result = code.join(';\n') + ';' + sourceCode
    }
  }

  // linaria transform
  let linaria
  try {
    linaria = require('linaria')
  } catch (e) {} // eslint-disable-line no-empty

  if (linaria) {
    let transformResult
    try {
      transformResult = transformLinaria({
        sourcePath: filePath,
        sourceCode: result
      })
    } catch (e) {
      console.error(e)
    }

    // linaria 转换的代码
    if (transformResult && transformResult.code) {
      result = transformResult.code
    }
  }

  return result
}
