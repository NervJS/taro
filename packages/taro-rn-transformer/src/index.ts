import * as path from 'path'
import appLoader, { getAppPages } from './app'
import pageLoader from './page'
import { TransformType, globalAny } from './types/index'
import { isPageFile, getCommonStyle } from './utils'

// eslint-disable-next-line import/no-commonjs
module.exports.transform = function ({ src, filename, options }: TransformType) {
  let code = src
  const sourceDir = options?.sourceRoot || 'src'
  const entryName = options?.entry || 'app'
  if (!(/node_modules/.test(filename)) && filename.indexOf(sourceDir) !== -1 && !filename.includes('.config')) {
    const appPath = path.join(options.projectRoot, sourceDir, entryName)
    const basePath = path.join(options.projectRoot, sourceDir)
    const pages = getAppPages(appPath)
    globalAny.__taroAppPages = pages.map(item => sourceDir + item)
    globalAny.__taroCommonStyle = getCommonStyle(appPath, basePath)
  }
  if (options.isEntryFile(filename)) {
    code = appLoader({
      filename: filename,
      projectRoot: options.projectRoot,
      sourceDir: sourceDir,
      appName: options.appName || 'taroDemo',
      entryName: entryName,
      designWidth: options?.designWidth || 750,
      deviceRatio: options?.deviceRatio || {
        640: 2.34 / 2,
        750: 1,
        828: 1.81 / 2
      }
    })
  } else if (isPageFile(filename, sourceDir)) {
    code = pageLoader({
      projectRoot: options.projectRoot,
      sourceCode: src,
      sourceDir: sourceDir,
      filename
    })
  }
  return options.nextTransformer({
    src: code,
    filename: filename,
    options: options
  })
}
