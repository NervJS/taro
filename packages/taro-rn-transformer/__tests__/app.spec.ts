import * as path from 'path'

import { transform } from '../src/index'

const App = `class HomeScreen extends React.Component {
  render () {
    return this.porps.children
  }
}`

function run () {
  const options = {
    projectRoot: path.resolve(__dirname),
    isEntryFile: () => { return true },
    nextTransformer: ({ src }) => { return src }
  }
  const code = transform({ src: App, filename: './src/app', options })
  return code
}

describe('app-loader', () => {
  it('app', () => {
    const result = run()
    expect(result).toMatchSnapshot()
  })
})
