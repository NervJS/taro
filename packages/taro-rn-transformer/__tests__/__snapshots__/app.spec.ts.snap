// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app-loader app 1`] = `
"import 'react-native-gesture-handler'
  import { AppRegistry } from 'react-native'
  import { createReactNativeApp, createPageConfig } from '@tarojs/runtime-rn'
  import Component from './src/app'
  import pagesIndexIndex from './src/pages/index/index';import pagesExpoIndex from './src/pages/expo/index';import pagesAppIndex from './src/pages/app/index';import pagesAppText from './src/pages/app/text';import package1PagesSubTest1 from './src/package1/pages/sub/test1';import package1PagesSubApp from './src/package1/pages/sub/app'
  import AppComponentConfig from './src/app.config';
  const pagesIndexIndexConfig = {};const pagesExpoIndexConfig = {};const pagesAppIndexConfig = {};const pagesAppTextConfig = {};const package1PagesSubTest1Config = {};const package1PagesSubAppConfig = {}

  AppComponentConfig.tabBar = {"backgroundColor":"#ffffff","color":"#999","selectedColor":"#ff552e","list":[{"pagePath":"pages/index/index","iconPath":"data:image/png;base64,","selectedIconPath":"data:image/png;base64,","text":"首页"},{"pagePath":"pages/expo/index","iconPath":"data:image/png;base64,","selectedIconPath":"data:image/png;base64,","text":"expo"}]}

  const buildConfig = {"designWidth":750,"deviceRatio":{"640":1.17,"750":1,"828":0.905}}
  const config = { appConfig: { ...buildConfig, ...AppComponentConfig } }
  global.__taroAppConfig = config
  config['pageList'] = [{name:'pagesIndexIndex',pagePath:'/pages/index/index',component:createPageConfig(pagesIndexIndex,{...pagesIndexIndexConfig,pagePath:'/pages/index/index'})},{name:'pagesExpoIndex',pagePath:'/pages/expo/index',component:createPageConfig(pagesExpoIndex,{...pagesExpoIndexConfig,pagePath:'/pages/expo/index'})},{name:'pagesAppIndex',pagePath:'/pages/app/index',component:createPageConfig(pagesAppIndex,{...pagesAppIndexConfig,pagePath:'/pages/app/index'})},{name:'pagesAppText',pagePath:'/pages/app/text',component:createPageConfig(pagesAppText,{...pagesAppTextConfig,pagePath:'/pages/app/text'})},{name:'package1PagesSubTest1',pagePath:'/package1/pages/sub/test1',component:createPageConfig(package1PagesSubTest1,{...package1PagesSubTest1Config,pagePath:'/package1/pages/sub/test1'})},{name:'package1PagesSubApp',pagePath:'/package1/pages/sub/app',component:createPageConfig(package1PagesSubApp,{...package1PagesSubAppConfig,pagePath:'/package1/pages/sub/app'})}]
  AppRegistry.registerComponent('taroDemo',() => createReactNativeApp(Component,config,createPageConfig(pagesIndexIndex,{...pagesIndexIndexConfig,pagePath:'/pages/index/index'})))
  "
`;
